File: desktop/macos/Sources/Localyze/Research/ReActAgent.swift
import Foundation

@MainActor
final class ReActAgent {
    enum StepKind: String { case plan, reason, act, observe, final }
    struct Step { let kind: StepKind; let text: String }

    private let backend: MLXBackend
    private let maxSteps = 12

    init(backend: MLXBackend) {
        self.backend = backend
    }

    var webSearchEnabled: Bool {
        get { SettingsStore.shared.webSearchEnabled }
        set { SettingsStore.shared.webSearchEnabled = newValue }
    }

    func run(userPrompt: String) -> AsyncThrowingStream<Step, Error> {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    let webOn = self.webSearchEnabled
                    let tools = Tools.shared.build(includeWeb: webOn)
                    let dispatch = Dictionary(uniqueKeysWithValues: tools.map { ($0.name, $0.call) })

                    var conversation = "<start_of_turn>system\n"
                        + Self.systemPrompt(webEnabled: webOn) + "<end_of_turn>\n"
                        + "<start_of_turn>user\n" + userPrompt + "<end_of_turn>\n"
                        + "<start_of_turn>model\n"

                    for step in 0..<self.maxSteps {
                        continuation.yield(Step(kind: .reason, text: "step \(step + 1)"))

                        var stepOutput = ""
                        for try await tok in self.backend.generate(prompt: conversation, maxTokens: 512) {
                            stepOutput += tok
                        }

                        if let finalText = Self.matchFinal(stepOutput) {
                            continuation.yield(Step(kind: .final, text: finalText))
                            continuation.finish()
                            return
                        }

                        guard let call = Self.matchToolCall(stepOutput) else {
                            // No tool call, no <final> — treat as direct answer.
                            continuation.yield(Step(kind: .final, text: stepOutput.trimmingCharacters(in: .whitespacesAndNewlines)))
                            continuation.finish()
                            return
                        }

                        continuation.yield(Step(kind: .act, text: "\(call.name) \(call.argsJson)"))

                        let result: [String: Any]
                        if let fn = dispatch[call.name] {
                            do {
                                result = try await fn(call.argsObj)
                            } catch {
                                result = ["error": error.localizedDescription]
                            }
                        } else {
                            result = ["error": "tool '\(call.name)' is not available"]
                        }

                        let resultJson = (try? JSONSerialization.data(withJSONObject: result))
                            .flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
                        continuation.yield(Step(kind: .observe, text: resultJson))

                        conversation += stepOutput
                            + "\n<tool_result name=\"\(call.name)\">\n"
                            + resultJson + "\n</tool_result>\n"
                    }

                    continuation.yield(Step(kind: .final, text: "[hit max steps]"))
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    // MARK: - parsing

    struct ToolCall { let name: String; let argsObj: [String: Any]; let argsJson: String }

    private static func matchFinal(_ text: String) -> String? {
        let pattern = #"<final>([\s\S]*?)</final>"#
        guard let rx = try? NSRegularExpression(pattern: pattern) else { return nil }
        let ns = text as NSString
        guard let m = rx.firstMatch(in: text, range: NSRange(location: 0, length: ns.length))
        else { return nil }
        return ns.substring(with: m.range(at: 1)).trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func matchToolCall(_ text: String) -> ToolCall? {
        let pattern = #"<tool\s+name="([^"]+)"\s*>([\s\S]*?)</tool>"#
        guard let rx = try? NSRegularExpression(pattern: pattern) else { return nil }
        let ns = text as NSString
        guard let m = rx.firstMatch(in: text, range: NSRange(location: 0, length: ns.length))
        else { return nil }
        let name = ns.substring(with: m.range(at: 1))
        let argsJson = ns.substring(with: m.range(at: 2)).trimmingCharacters(in: .whitespacesAndNewlines)
        let obj = (try? JSONSerialization.jsonObject(with: Data(argsJson.utf8))) as? [String: Any] ?? [:]
        return ToolCall(name: name, argsObj: obj, argsJson: argsJson)
    }

    private static func systemPrompt(webEnabled: Bool) -> String {
        var s =
            "You are Localyze, a private on-device assistant.\n\n"
            + "Tools (call by emitting <tool name=\"NAME\">{json args}</tool>):\n"
            + "  memory.search(query, limit=5)\n"
            + "  files.search(query, limit=5)\n"
            + "  calc(expr)\n"
            + "  run(lang, code)\n"
            + "  system.info()\n"
        if webEnabled { s += "  web.search(query, n=5)\n" }
        s += "\nFinalize with: <final>...answer...</final>\n"
           + "Cite tool results. Prefer 1-3 tool calls. Be concise.\n"
        return s
    }
}

File: desktop/windows/Localyze/Research/ReActAgent.cs
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using System.Threading;
using Localyze.Inference;

namespace Localyze.Research;

public sealed record ResearchStep(string Kind, string Text);

/// <summary>
/// Deep research mode loop. Reason → Act → Observe → repeat.
/// Web search is a single binary flag exposed via SettingsStore.
/// Tool-call wire format is documented in desktop/shared/research-agent.md.
/// </summary>
public sealed class ReActAgent
{
    private readonly OnnxBackend _backend;
    private readonly Tools _tools = new();
    private const int MaxSteps = 12;

    private static readonly Regex FinalRx =
        new(@"<final>([\s\S]*?)</final>", RegexOptions.Compiled);
    private static readonly Regex ToolRx =
        new(@"<tool\s+name=""([^""]+)""\s*>([\s\S]*?)</tool>", RegexOptions.Compiled);

    public bool WebSearchEnabled
    {
        get => SettingsStore.Instance.WebSearchEnabled;
        set => SettingsStore.Instance.WebSearchEnabled = value;
    }

    public ReActAgent(OnnxBackend backend) => _backend = backend;

    public async IAsyncEnumerable<ResearchStep> RunAsync(
        string userPrompt,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        var webOn = WebSearchEnabled;
        var available = _tools.Build(webOn);
        var dispatch = new Dictionary<string, ToolFunc>();
        foreach (var t in available) dispatch[t.Name] = t.Call;

        var conv = new StringBuilder()
            .Append("<start_of_turn>system\n").Append(SystemPrompt(webOn)).Append("<end_of_turn>\n")
            .Append("<start_of_turn>user\n").Append(userPrompt).Append("<end_of_turn>\n")
            .Append("<start_of_turn>model\n");

        for (var step = 0; step < MaxSteps; step++)
        {
            yield return new ResearchStep("reason", $"step {step + 1}");
            var sb = new StringBuilder();
            await foreach (var tok in _backend.GenerateAsync(conv.ToString(), 512, ct))
            {
                sb.Append(tok);
            }
            var stepText = sb.ToString();

            if (FinalRx.Match(stepText) is { Success: true } fm)
            {
                yield return new ResearchStep("final", fm.Groups[1].Value.Trim());
                yield break;
            }

            var tm = ToolRx.Match(stepText);
            if (!tm.Success)
            {
                yield return new ResearchStep("final", stepText.Trim());
                yield break;
            }

            var toolName = tm.Groups[1].Value;
            var argsJson = tm.Groups[2].Value.Trim();
            var argsObj  = (JsonObject?)JsonNode.Parse(string.IsNullOrEmpty(argsJson) ? "{}" : argsJson)
                           ?? new JsonObject();
            yield return new ResearchStep("act", $"{toolName} {argsJson}");

            JsonObject result;
            if (dispatch.TryGetValue(toolName, out var fn))
            {
                try { result = await fn(argsObj); }
                catch (System.Exception ex) { result = new JsonObject { ["error"] = ex.Message }; }
            }
            else
            {
                result = new JsonObject { ["error"] = $"tool '{toolName}' is not available" };
            }
            var resultJson = result.ToJsonString();
            yield return new ResearchStep("observe", resultJson);

            conv.Append(stepText)
                .Append("\n<tool_result name=\"").Append(toolName).Append("\">\n")
                .Append(resultJson)
                .Append("\n</tool_result>\n");
        }

        yield return new ResearchStep("final", "[hit max steps]");
    }

    private static string SystemPrompt(bool webEnabled)
    {
        var sb = new StringBuilder()
            .Append("You are Localyze, a private on-device assistant.\n\n")
            .Append("Tools (call by emitting <tool name=\"NAME\">{json args}</tool>):\n")
            .Append("  memory.search(query, limit=5)\n")
            .Append("  files.search(query, limit=5)\n")
            .Append("  calc(expr)\n")
            .Append("  run(lang, code)\n")
            .Append("  system.info()\n");
        if (webEnabled) sb.Append("  web.search(query, n=5)\n");
        sb.Append("\nFinalize with: <final>...answer...</final>\n")
          .Append("Cite tool results. Prefer 1-3 tool calls. Be concise.\n");
        return sb.ToString();
    }
}
